### PR TITLE
Codemon Fetch

### LINKED ISSUE #17: Scrape codeforces and fetch input output
Command : `codemon fetch` inside contest directory.
- Check if the current working directory is a contest directory ? If not, display error message to `cd` into a contest directory.
- Parse the contest number from the directory name, if contest number is not parsed, display error message to rename directory with the contest number i.e., if contest round is Codeforces Round 1234 then directory should be named `1234` or maybe `round1234`. Contest number should be present mandatorily.
- Alternately give user option to add contest name as `codemon fetch <contestName>` .
- After getting the contest number, go to Contest URL (https://codeforces.com/contest/1234/problems), scrape all inputs and outputs and store in respective question folder. For example, inputs and outputs of problem A should be stored in `A/a.in` and `A/a.out` respectively.

### PR DESCRIPTION
**Pull request by Enigmage**
Closes #17

#### Additions.

- `CodemonFetch` Module
  - Compatible with #12, #13 and #14 as the `make_structure` function in `CodemonFetch` module first checks if the necessary files and directories exist and then makes them if they don't. Once #12 #13 #14 are integrated into the project there would be no need to make any changes in this module as `make_structure`  function can then only as a safety check for the presence of Input and Output files.
  - `fetch_tests` function takes the contest name as argument and scrapes codeforces problem section to get the sample test cases and write them to input and output files (`A.in` and `A.op`  etc.) of respective problems.
  - if wrong contest name/number is provided, an error message is displayed.
- Dependencies
  - Added `requests` and `bs4` for scraping. 
  - Added `tqdm` for progress bar while fetching sample test cases.

#### Modifications
- `codemon.py`:
  - Added conditional to read fetch command. 
    ```py
        elif arg == "fetch":
        try:
          # if User provides contest number as arguement.
          fetch_tests(sys.argv[countArg])
        except IndexError:
          # Extract all numbers from current directory name.
          contest_number = ''.join(re.findall(r'\d+', os.getcwd().split('/')[-1]))
          fetch_tests(contest_number)
    ```
  - if user provides contest number as argument it will be simply passed to `fetch_tests` function. Otherwise there will be an
IndexError. In that case the user hasn't provided an argument. So, the contest number will be parsed from the current contest directory. Here, number of the contest is extracted using regex. Assumption is that the correct contest number should be present in the contest directory name. 
 
- `CodemonMeta` : 
  - A new way of extracting filenames was added to solve an inconsistency with scraping sample test cases. For example, contest number `1234` has problems `A, B1, B2, C, D, E, F`. So by the old way, sample test cases for problem B2 might get stored under the name of `C`. Or, in other cases where there are problems till `G` the sample test cases for the last problem will be missing.  
  - So to create a consistent interface for the user, file names will be extracted by scraping the contest page. Again, same assumption applies, the contest number should be present in the contest name provided by the user. If contest number is not there or if it is wrong, a generic `A, B, C, D, E, F` is used for filenames.

### CODE REVIEW

**Comment by ankushbhardwxj** [comment]
@Enigmage - I love this idea of scraping the questions from Codeforces and creating the files accordingly. I haven't reviewed your code yet but I'd be amazed if you could extend your functionality to create the directory structure described in #48 , and create the files based on flag `-java`, `-py`, `-cpp` (default, if no flag then cpp files). In that case, I can award you the points of all three issues along with bonus points for coming up with this idea.

**Comment by ankushbhardwxj** [comment]
Check my [comment](https://github.com/ankingcodes/codemon/pull/30#discussion_r533976074) in the review for a better way to parse arguments from the shell. The `isCppFile` variable should be originally be `True` since if no flag is given, then we directly create Cpp files.,

**Comment by ankushbhardwxj** [comment]
Did some review and testing, you need to make the following changes:
- `codemon fetch` is serving two purposes, i.e., somewhat replaces `codemon init`, so I suggest adding a flag `-f` and `--fetch` along with `codemon init`, written as `codemon init --fetch 1455`, which would fetch all file names and testcases and create the recommended files. If only `codemon init` is used, create default directories (A,B, ..., G). 
- If user adds flag `-py` , `-java` or `-cpp` then create files with extension and template corresponding to them which will be declared in Codemon Meta. If no flag is given, proceed with C++ files. 
- Add verbose messages, such as `Creating file {somename}`, `Creating directories for contest`, `Fetching inputs and outputs` etc.

**Comment by Enigmage** [comment]
@ankingcodes  thanks for the review. 
> I'd be amazed if you could extend your functionality to create the directory structure described in #48 , and create the files based on flag -java, -py, -cpp (default, if no flag then cpp files)

Yes I would love to add that. I actually wanted to but since there were conflicting PRs #48 and #50 I thought i'll have to make a separate PR for extending this functionality once those changes are merged. But if you want I can extend that functionality in this PR itself. I have a very good  generic solution in mind that will be able to initialize directory structure for all three supported languages with very little code addition, It should pretty much solve #12 #13 and #14 and  of course the original issue to this PR. Still, if you plan to merge #48 and #50 before this, just ping me because then I'll have to make my additions over the changes made by those PRs, to extend this functionality.

>Check my comment in the review for a better way to parse arguments from the shell. The isCppFile variable should be originally be True since if no flag is given, then we directly create Cpp files.,

Sure, will follow this.

>  so I suggest adding a flag -f and --fetch along with codemon init, written as codemon init --fetch 1455, which would fetch all file names and testcases and create the recommended files. 

That's a great Idea, this way user has won't have to enter two separate commands and can get the sample test cases and initialize contest directory in a single command.

> codemon fetch is serving two purposes, i.e., somewhat replaces codemon init

No actually right now it looks like `codemon fetch` its replacing `init` since changes for #12 #13 and #14 haven't been merged into the project and `init` only initializes 6 source files with an input.txt and not the necessary question folders and input/output files that `codemon fetch` depends on. So, this is why I have added checks in `codemon fetch`. If you look at `make_structure` function in `CodemonFetch` module, it first checks whether the question folders and input/output files exist or not. If they do, it won't create them and just simply proceed to write test cases to them. This way once changes for #12, #13 and #14 are merged in `codemon init` ,  `codemon fetch` won't replace it but will act as an extra safety check to the functionality of `codemon init`. This is what I meant when I said it is compatible with other three issues of initialization.

>If only codemon init is used, create default directories (A,B, ..., G).

I have a suggestion for this one. We can create the default directory structure( A,B,....,G) without scraping only when the contest number provided by the user is wrong and doesn't coincide with any real contest on CF. I have already added error checks for this in `codemon fetch` where user provides wrong contest number. Otherwise, if we make the default directory structure with `codemon init` without validating the contest name and user runs `codemon fetch` inside this directory then it would again create the inconsistency that I was talking about earlier and would affect the functionality of `codemon fetch` .  What do you think?

> If user adds flag -py , -java or -cpp then create files with extension and template corresponding to them which will be declared in Codemon Meta. If no flag is given, proceed with C++ files.
Add verbose messages, such as Creating file {somename}, Creating directories for contest, Fetching inputs and outputs etc.

Okay will take care of this.

**Comment by alphaX86** [comment]
@Enigmage does it resolve issues without merging other templates or do I need to change any of them in my PR before merging yours? I'm asking this doubt to avoid conflict in mine (#50 ) and in other branches. Thanks

**Comment by Enigmage** [comment]
@alphaX86 if @ankingcodes merges your PR first, you won't have have to change anything, I will make the necessary additions over your changes to add my functionality. This is why I have asked which ones he is planning to merge first. Let's see what @ankingcodes has to say.

**Comment by ankushbhardwxj** [comment]
>  I have a very good generic solution in mind that will be able to initialize directory structure for all three supported languages with very little code addition, It should pretty much solve #12 #13 and #14 and of course the original issue to this PR

@Enigmage - I'm a fan of less code doing more things. Your PR is already doing that quite creatively. I don't think it will be a problem to extend the functionality of this PR to what #50 #48 are trying to do. 

You need not worry about the merging of other PRs. If you have a better idea, you should always present it in your PR. It's the responsibility of the maintainer to access other PRs and merge the best one. For example, other PRs are right now doing the same thing you're trying to do, but you have a solution which is better and requires less code than those PRs (I had the same thing in my mind too, was expecting someone to get it done). So go ahead and implement those. 

Just to clarify things, I'd reiterate the solution again. 
- `codemon init` (without `-f` or `--fetch`) : create directory structure normally (A,B,C...,G), default in C++ else Python or Java corresponding to the flags `-py` or `-java` given by user as arguments.
- `codemon init --fetch` or `codemon init -f`: create directory structure by scraping data from codeforces. Remember to create test case files. 
- (optional) `codemon init -n `: create a single file with given name, if possible parse the problem and contest number (eg: `1455B` should have problem name B and contest number 1455) and if user gives flag `--fetch` or `-f` then fetch the input output and create the necessary files. 
- (optional) Create an issue for scraping inputs outputs from codechef website, google codejam website, kickstart website, AtCoder website etc.

**Comment by Enigmage** [comment]
Okay cool. I will then add the required changes and ping you for review asap.

**Comment by Enigmage** [comment]
Hey @ankingcodes can you confirm what exactly the directory structure should look like after running init. I referred to your comments on various issues and PRs but its still not completely confirmed. This is what I'm getting at:
```
contest_name/
  A /
    A.cpp
    A.in
    A.op
    test_case.cpp
    test_case
  B /
......
```
If there is anything missing or wrong do point out. I'm done with parsing part and almost done with `init` just needed a confirmation of the final picture you had in mind. Thanks.

**Comment by ankushbhardwxj** [comment]
@Enigmage - I had discussed about having testcase files inside each directory, which would be cleaned up using `codemon clean`, but on second thoughts, I think that would be another effort for the user and the concept behind codemon is to keep user effort as minimum as possible, so as a result, I believe we should get rid of it from inside the problem directory and place it on the contest directory. That way even if a user doesn't run `codemon clean`, the amount of useless testcase files gets reduced.

**Comment by Enigmage** [comment]
@ankingcodes thanks, got it.

**Comment by ankushbhardwxj** [comment]
@Enigmage - resolve the conflicts 👀

**Comment by Enigmage** [comment]
@ankingcodes Done.

File: setup.py
Comment (ankushbhardwxj) [inline setup.py]: getting an error in installing, you missed a `,` here

**Review by ankushbhardwxj — CHANGES_REQUESTED** [review]

**Comment by ankushbhardwxj** [comment]
@Enigmage - package installation fails

**Comment by ankushbhardwxj** [comment]
@Enigmage - I think you made some mistake while resolving conflicts. I dont see any of your latest commits of the changes I had requested.
Look into it.

**Comment by Enigmage** [comment]
@ankingcodes I just made the commits that resolved the conflict. I will be making the changes you requested soon, working on them on a different branch.
>getting an error in installing, you missed a , here

yes you are right I'll resolve it wait.

**Comment by Enigmage** [comment]
@ankingcodes resolved the issue, installation should be working now. I will be making the requested commits soon.

**Comment by ankushbhardwxj** [comment]
@Enigmage - Ping me for review once you complete the features mentioned in https://github.com/ankingcodes/codemon/pull/55#issuecomment-739752918

**Comment by Enigmage** [comment]
@ankingcodes  So the changes are done. You can review now.

Closes #12 #13 #14 and #17

### What I didn't add/What needs to be done separately:

- Default templates for java and python. So that others who are interested can add them. I'll open issues with instructions on how to add templates for the two. 
- Listen behaviour. Since the directory structure is new and two new languages are added, work will be needed in the listening behaviour of codemon, since that's a separate issue, I haven't made a lot of changes there. But I have added changes such that listen works for cpp files, takes input from `test_case` and compiles. We'll have to open issues for adding listening behaviour for Python and Java Files.
- I haven't added the third optional task you gave me about adding the fetch functionality to single files. With your approval, I would like to open a separate issue for it and work on it separately.


### What I added
I'll document my changes over three parts I divided my work into.

#### Parsing:
 
- Borrowed your Boolean approach. So there are three parts to your approach: first initializing the booleans, second setting their values based on command line arguments provided and third performing operations based on the booleans. So I wrapped the first two parts of your approach into a class `Parser`, booleans being the data members and a member function `parse` that takes `sys.argv` as argument and sets the values of those booleans. I added this class in a separate module so that the code is readable, modular and the `main` function only contains the third part of your approach. Also makes it easy for developers in the future to add new commands.
  - I have separated flags from non flag arguments in the `Parser` class so that flags don't have to be fixed at a certain position and for separation of concerns. 

#### Initialization

- Initialization support for for all three languages based on flags provided by the user. Didn't need to add any new function for this, just had to make the present `init` function accept a third argument `init_flags` which is a dictionary of booleans containing different options set during the parse phase. Based on that `init` is done for all three filetypes. Init for single files also added based on flags, same for practice files. 

#### Fetch 
- Added a `-f/--fetch` option with `codemon init` for initializing directory and fetching sample test cases at the same time.
- Rest of the functionality is same.

**Comment by Enigmage** [comment]
@ankingcodes  if you're reviewing this, I have added a few more changes, error checks basically.

- Earlier the user could get away with typing commands like `codemon reg this xx` or `codemon listen abc` or ` codemon -py -java init` etc and there would be no errors. I have added error checks that will show help section if such faulty commands are entered.
  - If user enters two language flags at the same time, it will cause an error and show help section
  - if user enters extra arguments in commands like `listen` , `reg` that take only one argument, it will cause an error and show help section.

- In case the user did not have internet connection `fetch` would error out and contest directory won't be created. Now, I have added exception handling which in such a situation will create a default directory structure, so the user won't have to depend on internet connection for initialization of contest directory.

Also some code refactoring.

**Comment by ankushbhardwxj** [comment]
@Enigmage - Reviewed your PR. Excellent work ! Everything seems to be on point. Many thanks for refactoring code and going OOP style. There are a few tweaks I'd like to be made but I'm sure I can do that myself or maybe make separate issues for them. Thanks for contributing.

**Pull request merged**
